### PR TITLE
fasttext Classifier

### PR DESCRIPTION
**Pull request by mathieumb**
Very fast text classification.
We should eventually recode the math with eigen and stop linking against the fasttext fork altogether, since its not a lot of code.
It would also be nice to be able to output the embedding (the "input" matrix).

### CODE REVIEW

File: ml/jml/fasttext_classifier.cc [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+FastTest_Classifier::
+~FastTest_Classifier()
+{
+}
+    
+void
+FastTest_Classifier::
+swap(FastTest_Classifier & other)
+{
+    Classifier_Impl::swap(other);
```
Comment (guyd) [inline ml/jml/fasttext_classifier.cc]: Why this call if it id not implemented?

File: ml/jml/fasttext_generator.cc [hunk 2]
```diff
@@ -0,0 +1,236 @@
+/* fasttext_tree_generator.cc                                       -*- C++ -*-
+   Mathieu Marquis Bolduc, 2 March 2027
```
Comment (guyd) [inline ml/jml/fasttext_generator.cc]: 2027 -> 2017!

File: ml/jml/fasttext_generator.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
+/* fasttext_tree_generator.h                                       -*- C++ -*-
+   Mathieu Marquis Bolduc, 2 March 2027
```
Comment (guyd) [inline ml/jml/fasttext_generator.h]: 2027 -> 2017

**Comment by guyd** [comment]
My understanding of fasttext is limited so you may want to get a review from another party as well.  The code looks OK.  The test coverage would benefit of a better/more convincing test.

Made very minor comments.
+1

**Comment by mathieumb** [comment]
I have a better test but the model is 200 meg and the dataset 800 meg its a bit heavy for unit tests. fmaillet will also review.

File: ml/jml/fasttext_classifier.cc (hunk 1 above)
Comment (mathieumb) [inline ml/jml/fasttext_classifier.cc]: Well we're going to need that call regardless when/if we implement it, so its to not forget it :)

File: container_files/public_html/doc/builtin/ClassifierConf.md [hunk 4]
```diff
@@ -132,6 +133,15 @@ all the values in two, using the threshold that provides the best separation
 of classes. You can always do you own discretization, for instance using a
 `CASE` expression.
 
+<a name="fasttext"></a>
+### FastText (type=fasttext)
+
+![](%%jmlclassifier fasttext)
+
+Note that our version of the Fast Text Classifier only supports feature counts, and currently
```
Comment (mailletf) [inline container_files/public_html/doc/builtin/ClassifierConf.md]: tu veux dire quoi par "feature counts" ?

File: ml/jml/fasttext_generator.cc (hunk 2 above)
Comment (mailletf) [inline ml/jml/fasttext_generator.cc]: back to the future!!!

File: ml/jml/fasttext_generator.cc [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+#include <random>
+#include <mutex>
+
+
+using namespace std;
+
+namespace ML {
+
+/*****************************************************************************/
+/* DECISION_TREE_GENERATOR                                                   */
```
Comment (mailletf) [inline ml/jml/fasttext_generator.cc]: wrong name?

File: ml/jml/fasttext_classifier.cc (hunk 1 above)
Comment (FinchPowers) [inline ml/jml/fasttext_classifier.cc]: Is there a possibility that an end user catch this exception? Is so, we should implement everything that's needed.

File: ml/jml/fasttext_generator.h [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+#include "fasttext_classifier.h"
+
+namespace ML {
+
+
+/*****************************************************************************/
+/* FastText_Generator                                                        */
+/*****************************************************************************/
+
+/** Class to generate a classifier.  The meta-algorithms (bagging, boosting,
```
Comment (mailletf) [inline ml/jml/fasttext_generator.h]: commentaire est bon?

File: testing/fasttext_test.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
+# -*- coding: utf-8 -*-
+
+# ##
+# Mathieu Marquis Bolduc, March 2nd 2017
+# This file is part of MLDB. Copyright 2017 mldb.ai inc. All rights reserved.
+# ##
+
+mldb = mldb_wrapper.wrap(mldb) # noqa
+
+class FastTextTest(MldbUnitTest):
```
Comment (mailletf) [inline testing/fasttext_test.py]: un test pour que ça plante avec une erreur claire si on essaye avec regression ou categorical?

File: ml/jml/fasttext_classifier.cc [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+        store << matrix->m_;
+        store << matrix->n_;
+        fasttext::real* p = matrix->data_;
+        for (int i = 0; i < matrix->m_*matrix->n_; ++i, p++) {
+            store << *p;
+            ExcAssert(!std::isnan(*p));
+        }
+    };
+
+    string id = class_id();
```
Comment (FinchPowers) [inline ml/jml/fasttext_classifier.cc]: unused?

File: ml/jml/fasttext_classifier.cc [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+class_id() const
+{
+    return "FASTTEXT";
+}
+
+FastTest_Classifier *
+FastTest_Classifier::
+make_copy() const
+{
+    throw Exception("FastTest_Classifier make_copy: not tested");
```
Comment (FinchPowers) [inline ml/jml/fasttext_classifier.cc]: ❓

File: ml/jml/fasttext_generator.cc (hunk 2 above)
Comment (FinchPowers) [inline ml/jml/fasttext_generator.cc]: To the future and beyond!

File: ml/jml/fasttext_generator.cc [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+
+    args_->model = fasttext::model_name::sup;
+    args_->thread = training_data.example_count() < numCpus() ? 1 : numCpus();
+    args_->bucket = 0;
+    args_->dim = dims;
+    args_->epoch = epoch;
+    args_->verbose = verbose;
+
+
+    std::shared_ptr<fasttext::Matrix> input_ = std::make_shared<fasttext::Matrix>(features.size()+args_->bucket, args_->dim);
```
Comment (FinchPowers) [inline ml/jml/fasttext_generator.cc]: line too long

File: ml/jml/fasttext_generator.cc [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+    output_->zero();
+    fastTextModel.input_ = input_;
+    fastTextModel.output_ = output_;
+
+    //Note that we skip Fasttext's dictionary because the strings are already abstracted in our features  
+
+    std::atomic<int64_t> tokenCount(0);
+    auto trainThread = [&] (size_t threadId) {
+
+      int32_t start = threadId * training_data.example_count() / args_->thread;
```
Comment (FinchPowers) [inline ml/jml/fasttext_generator.cc]: Can `args_->thread == 0`?

File: ml/jml/fasttext_generator.cc [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+    
+      int64_t localTokenCount = 0;
+      std::vector<int32_t> line, labels;
+      while (tokenCount < args_->epoch * ntokens) {
+
+        //if we reach the end we wrap around
+        if (start >= training_data.example_count())
+            start = 0;
+
+        fasttext::real progress = fasttext::real(tokenCount) / (args_->epoch * ntokens);
```
Comment (FinchPowers) [inline ml/jml/fasttext_generator.cc]: Divide by zero check?

File: ml/jml/fasttext_generator.cc [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+            else {
+                auto featureit = output->featureMap.find(feature.first);
+                if (featureit != output->featureMap.end()) {
+                    size_t f = (*featureit);
+                    for (int i = 0; i < feature.second; ++i)
+                        line.push_back(f);        
+                }
+            }
+            localTokenCount++;
+            ++it;
```
Comment (FinchPowers) [inline ml/jml/fasttext_generator.cc]: It's the prefix/postfix increment valse?

File: ml/jml/fasttext_generator.cc [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+        }       
+
+        start++;        
+        if (labels.size() != 0 && line.size() > 0)
+        {
```
Comment (FinchPowers) [inline ml/jml/fasttext_generator.cc]: { on if line

File: ml/jml/fasttext_generator.cc [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+            std::uniform_int_distribution<> uniform(0, labels.size() - 1);
+            int32_t i = uniform(model.rng); 
+            model.update(line, labels[i], lr);
+        }
+     
+        if (localTokenCount > args_->lrUpdateRate) {
+          tokenCount += localTokenCount;
+          localTokenCount = 0;
+          if (threadId == 0 && args_->verbose > 1) {
+            fastTextModel.printInfo(progress, model.getLoss());
```
Comment (FinchPowers) [inline ml/jml/fasttext_generator.cc]: bad indentation

File: ml/jml/fasttext_generator.cc (hunk 11 above)
Comment (mathieumb) [inline ml/jml/fasttext_generator.cc]: Only if the machine has zero CPU, in which case we have more pressing issues :P

File: ml/jml/fasttext_generator.cc (hunk 13 above)
Comment (mathieumb) [inline ml/jml/fasttext_generator.cc]: it is an iterator so using prefix is slightly more efficient. localTokenCount is an atomic so it doesnt matter.

File: ml/jml/fasttext_generator.cc (hunk 12 above)
Comment (mathieumb) [inline ml/jml/fasttext_generator.cc]: if args_->epoch * ntokens == 0 then we cannot enter this block per virtue of the 'while' above

File: container_files/public_html/doc/builtin/ClassifierConf.md (hunk 4 above)
Comment (mathieumb) [inline container_files/public_html/doc/builtin/ClassifierConf.md]: Je veux dire que FastText prend des bag of words. La valeur numérique de la feature sera toujours interprétée comme le nombre de fois que la feature apparait dans l'exemple.

**Comment by mathieumb** [comment]
@guyd @mailletf @FinchPowers any blocker still?

File: ml/jml/fasttext_classifier.cc (hunk 1 above)
Comment (FinchPowers) [inline ml/jml/fasttext_classifier.cc]: @mathieumb ping

File: testing/fasttext_test.py (hunk 7 above)
Comment (FinchPowers) [inline testing/fasttext_test.py]: @mathieumb ping

**Comment by guyd** [comment]
ok for me.

File: ml/jml/fasttext_classifier.cc (hunk 1 above)
Comment (mathieumb) [inline ml/jml/fasttext_classifier.cc]: I couldnt find any case of classifiers being swapped in the codebase.

File: testing/fasttext_test.py (hunk 7 above)
Comment (mathieumb) [inline testing/fasttext_test.py]: Done

**Pull request merged**